include/bluetooth/bluetooth.h

```diff
@@ -398,6 +463,118 @@ int bt_le_adv_update_data(const struct bt_data *ad, size_t ad_len,
+/** @brief Increment an advertising set`s reference count.
+ *
+ *  Increment the reference count of an advertising set object.
+ *
+ *  @param adv Advertising set object.
+ *
+ *  @return Advertising set object with incremented reference count.
+ */
+struct bt_adv *bt_adv_ref(struct bt_adv *adv);
+
+/** @brief Decrement an advertising set's reference count.
+ *
+ *  Decrement the reference count of an advertising setobject.
+ *
+ *  @param adv Advertising set object.
+ */
+void bt_adv_unref(struct bt_adv *adv);
+
+/** @brief Create advertising set.
+ *
+ *  Create a new advertising set and set advertising parameters.
+ *
+ *  @param param Advertising parameters.
+ *
+ *  @return Zero on success or (negative) error code otherwise.
+ *  @return -ECONNREFUSED When connectable advertising is requested and there
+ *			  is already maximum number of connections established.
+ *			  This error code is only guaranteed when using Zephyr
+ *			  controller, for other controllers code returned in
+ *			  this case may be -EIO.
+ */
+struct bt_adv *bt_le_adv_set_create(const struct bt_le_adv_param *param);
+
+/** Start advertising with the given advertising set
+ *
+ *  Start advertising with a specific advertising set.
+ *  If both timeout and advertising event limit is given then the advertising
+ *  set will be stopped once one of them is reached.
+ *
+ *  @param adv         Advertising set object.
+ *  @param timeout     Timeout in ms, K_FOREVER or K_NO_WAIT will disable timeout.
+ *  @param num_events  Stop advertising after the specified number of advertising
+ *                     events has been reached.
+ */
+int bt_le_adv_set_start(const struct bt_adv *adv, s32_t timeout,
+			u8_t num_events);
+
+/** Stop advertising with the given advertising set
+ *
+ *  Stop advertising with a specific advertising set.
+ *
+ *  @param adv Advertising set object.
+ *
+ *  @return Zero on success or (negative) error code otherwise.
+ */
+int bt_le_adv_set_stop(const struct bt_adv *adv);
+
+/** Set advertising set advertising or scan response data.
+ *
+ *  Set advertisement data or scan response data. If the advertising set is
+ *  currently advertising then the advertising data will be updated in
+ *  subsequent advertising events.
+ *
+ *  If the advertising set has been configured with Primary advertising then the
+ *  maximum data length is 31 bytes. If the advertising set has been configured
+ *  for extended advertising data length, then the maximum data length defined
+ *  by the controller.
+ *
+ *  NOTE: Not all scanners support extended data length advertising data.
+ *
+ *  @param ad      Data to be used in advertisement packets.
+ *  @param ad_len  Number of elements in ad
+ *  @param sd      Data to be used in scan response packets.
+ *  @param sd_len  Number of elements in sd
+ *
+ *  @return Zero on success or (negative) error code otherwise.
+ */
+int bt_le_adv_set_data(const struct bt_adv *adv,
+		       const struct bt_data *ad, size_t ad_len,
```

Comment:
Hmm I don't think flat buffers are suitable for EA, data can be up to 1650 bytes...

Reply:
Which data-structure would be better? `net_buf`?

Reply:
It's not quite a flat buffer in this case though since `ad` is an array  (one element for each data type to be included to the advertising data (the `ad_len` parameter is the array length). That said, net_buf might indeed be a more lucrative choice here.

Reply:
Right. I keep forgetting how that works. :)

Then I don't necessarily see the downside in continuing to use the same parameters as before.

What might be more important is making is the scan callback `recv` I think right now it has a `net_buf_simple`. And possibly `bt_data_parse`